spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/AsyncRabbitTemplate.java

```diff
@@ -0,0 +1,589 @@
+	}
+
+	@Override
+	public synchronized void stop() {
+		if (this.running) {
+			this.container.stop();
+		}
+		this.running = false;
+	}
+
+	@Override
+	public boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public int getPhase() {
+		return this.phase;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public void onMessage(Message message) {
+		String correlationId = new String(message.getMessageProperties().getCorrelationId(), this.charset);
+		if (correlationId != null) {
+			if (logger.isDebugEnabled()) {
+				logger.debug("onMessage: " + message);
+			}
+			RabbitFuture<?> future = this.pending.remove(correlationId);
+			if (future != null) {
+				if (future instanceof AsyncRabbitTemplate.RabbitConverterFuture) {
+					C converted = (C) this.template.getMessageConverter().fromMessage(message);
+					((RabbitConverterFuture) future).set(converted);
+				}
+				else {
+					((RabbitMessageFuture) future).set(message);
+				}
+			}
+		}
+	}
+
+	@Override
+	public void returnedMessage(Message message, int replyCode, String replyText, String exchange, String routingKey) {
+		String correlationId = new String(message.getMessageProperties().getCorrelationId(), this.charset);
+		if (correlationId != null) {
+			@SuppressWarnings("unchecked")
+			RabbitFuture<?> future = this.pending.remove(correlationId);
+			if (future != null) {
+				future.setException(new AmqpMessageReturnedException("Message returned", message, replyCode, replyText,
+						exchange, routingKey));
+			}
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public void confirm(CorrelationData correlationData, boolean ack, String cause) {
+		if (logger.isDebugEnabled()) {
+			logger.debug("Confirm: " + correlationData + ", ack=" + ack
+					+ (cause == null ? "" : (", cause: " + cause)));
+		}
+		String correlationId = correlationData.getId();
+		if (correlationId != null) {
+			RabbitFuture<?> future = this.pending.get(correlationId);
+			if (future instanceof SettableListenableFuture) {
+				future.setNackCause(cause);
+				((SettableListenableFuture<Boolean>) future.getConfirm()).set(ack);
+			}
+			else {
+				if (logger.isWarnEnabled()) {
+					logger.warn("Confirm: " + correlationData + ", ack=" + ack
+							+ (cause == null ? "" : (", cause: " + cause))
+							+ " no pending future - either canceled or the reply is already received");
+				}
+			}
+		}
+	}
+
+	private String getOrSetCorrelationIdAndSetReplyTo(Message message) {
+		String correlationId;
+		byte[] currentCorrelationId = message.getMessageProperties().getCorrelationId();
+		if (currentCorrelationId == null) {
+			correlationId = UUID.randomUUID().toString();
+			message.getMessageProperties().setCorrelationId(correlationId.getBytes(this.charset));
+			message.getMessageProperties().setReplyTo(this.replyAddress);
```

Comment:
So, are we OK do not set `replyTo` when we have already `currentCorrelationId` ?